[FLINK-30662][table] Planner supports delete

luoyuxia opened the pull request:
<!--
*Thank you very much for contributing to Apache Flink - we are happy that you want to help us improve Flink. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

*Please understand that we do not do this to make contributions to Flink a hassle. In order to uphold a high standard of quality for code contributions, while at the same time managing a large number of contributions, we need contributors to prepare the contributions well, and give reviewers enough contextual information for the review. Please also understand that contributions that do not follow this guide will take longer to review and thus typically be picked up with lower priority by the community.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix] [docs] Fix typo in event time introduction` or `[hotfix] [javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can set up Azure Pipelines CI to do that following [this guide](https://cwiki.apache.org/confluence/display/FLINK/Azure+Pipelines#AzurePipelines-Tutorial:SettingupAzurePipelinesforaforkoftheFlinkrepository).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.


**(The sections below can be removed for hotfixes of typos)**
-->

## What is the purpose of the change
To make planner supports delete including delete push down and row-level delete


## Brief change log
It contains two commits:
- Planner supports delete push down
  - Convert the `SqlDelete` to `LogicalTableModify`, then get the `DynamicTableSInk` for the table to be deleted.
  - Extract  the predicate in `Filter` contained in `LogicalTableModify` and resolve it to a list of `ResolvedExpression`
  - if the `DynamicTableSInk`is instance of `SupportsDeletePushDown`, call method `SupportsDeletePushDown#applyDeleteFilters` to apply filters, if it returns true, create a `DeleteFromFilterOperation`.
  - Call method `DeleteFromFilterOperation#getSupportsDeletePushDown#executeDeletion` to do the deletion in `TableEnvironmentImpl#executeInternal`  if the operation is `DeleteFromFilterOperation`.
 
- Planner supports row-level delete
  - Add a field `isDelete` in `SinkModifyOperation` to represents it's for delete.
  - If the delete push down is not applicable for the cases where the table sink doesn't implement  `SupportsDeletePushDown` or `SupportsDeletePushDown#applyDeleteFilters`, return  a `SinkModifyOperation` with `isDelete=true` and the `LogicalTableModify` converted from `SqlDelete`. 
  - Rewrite the `LogicalTableModify` to a RelNode that insert into the sink with rows
    - If the delete mode is `REMAINING_ROWS`, negate the predicate in `Filter` node.
    - Get the required columns by calling method `SupportsRowLevelDelete#applyRowLevelDelete`. We may rewrite the `LogicaTableScan` to add extra meta cols to the underlying table. Then create a `Project` node that only select the  required columns.
   - If  the delete mode is `DELETED_ROWS`, add a operator to set the `RowKind` to `RowKind#Delete` in `CommonExecSink`


## Verifying this change
- The test in `SqlToOperationConverterTest#testDelete` to verify delete statement will be converted to `DeleteFromFilterOperation` if delete can be pushed down or `SinkModifyOperation`.
- The test in `RowLevelDeleteTest` to verify the plan for row-level delete.
- The IT case in `DeleteTableITCase` to verify delete can be performed normally.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / **not documented**)

thexiay commented on flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/operations/DeletePushDownUtils.java:
```diff
@@ -0,0 +1,298 @@
+     * Get the {@link DynamicTableSink} for the table to be modified. Return Optional.empty() if it
+     * can't get the {@link DynamicTableSink}.
+     */
+    public static Optional<DynamicTableSink> getDynamicTableSink(
```
Is it copied from PlannerBase#getTableSink?

luoyuxia replied:
Just, most if from PlannerBase#getTableSink.

thexiay commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/TableEnvironmentImpl.java:
```diff
@@ -1400,6 +1411,23 @@ public TableResultInternal executeInternal(Operation operation) {
+            if (isStreamingMode) {
+                throw new TableException("DELETE TABLE is not supported for streaming mode now.");
+            }
+            Optional<Long> rows =
```
This implement has some limit:
1. Predicate cannot contains subquery, it must be a deterministic expressions.
2. This delete or update runtime execution is at client environment, maybe it's network environment has a littele bit different from taskmanager in cluster.(e.g. taskmanager can connect data source, but client cannot connect it , so execute in here may fail) 

Do you think those limitations are acceptable or not?

luoyuxia replied:
Thanks for attention.
Yes, it's acceptable.
1: When the predicate contain subquery, the delete can't be pushed,  it can fall back to SupportsRowLevelDelete which support to delete for all cases.
2: Still, only delete can pushed to table source, will we do the deletion in here. If the delete can be pushed down, it's be more like it's to drop some partition/delete files for storage based on files.  I think it's similar to drop partition which is also done in TableEnvironmentImpl.  For other storage like mysql, it expect to delegate the delete stamement to the underlying engine,  I think it's fine that client should have the ability to connect to the underlying engine.
Also, I think it's beneficial that execute it in client direclty which we also reduce the cost of generate/deploy job.

For more detail, you can refer to the [FLIP-282](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=235838061)

lincoln-lil commented on flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/runtime/batch/sql/DeleteTableITCase.java:
```diff
@@ -0,0 +1,89 @@
+        assertThat(rows).isEmpty();
+
+        // should throw exception for non-empty filter
+        assertThatThrownBy(() -> tEnv().executeSql("DELETE FROM t where a = 1"))
```
Can we extend current test utility to also support verifing the case that a sink can accept non-empty filters?

luoyuxia replied:
Yes, we can. But I think there's no need for in here I'm intended to test the case that although the table support delete pushdown, but it can't accept the filter, it should throw exception if it haven't implemented `SupportsRowLevelDelete` interface.

lincoln-lil replied:
If I haven't missed it, we currently have no such test cases？

and some thoughts for adding this case: for the delete scenarios, users may delete part of the data more offen? If they want to delete the whole table, I think `truncate` maybe the better choice.

luoyuxia replied:
I see. Yes, we need to cover the more common use scenario. I'll update it then.

lincoln-lil commented on flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/operations/SqlToOperationConverter.java:
```diff
@@ -1492,6 +1500,41 @@ private Operation convertStopJob(SqlStopJob sqlStopJob) {
+                    return new DeleteFromFilterOperation(supportsDeletePushDown, filters.get());
+                }
+            }
+        }
```
should we raise an error when the returned dynamicTableSink is empty？(test case also needed)

luoyuxia replied:
If the table is deprecated `TableSInk`, it will get empty. 
I think we can raise an error in `PlannerBase#translateToRel(modifyOperation: ModifyOperation)`,  when the sink is 
a deprecated `TableSInk`, we then throw exception.  It'll be more unified for we also need to do such check for row-level delete/udpate.
If that's the case, I will add the code while I'm fixing the second commit(support row-level delete).

lincoln-lil replied:
the unified validation sounds good to me

lincoln-lil commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/DeleteFromFilterOperation.java:
```diff
@@ -0,0 +1,56 @@
+/** The operation for deleting data in a table according to filters directly. */
+@Internal
+public class DeleteFromFilterOperation implements Operation {
+    private final SupportsDeletePushDown supportsDeletePushDown;
```
-> 'supportsDeletePushDownSink' maybe more clear

lincoln-lil commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/DeleteFromFilterOperation.java:
```diff
@@ -0,0 +1,56 @@
+    private final List<ResolvedExpression> filters;
+
+    public DeleteFromFilterOperation(
+            SupportsDeletePushDown supportsDeletePushDown, List<ResolvedExpression> filters) {
```
add `Nonnull` annotation and null check since only accept non-empty objects when do conversion

lincoln-lil reviewed (commented):
@luoyuxia thanks for your contribution! 
For the 1st commit which supports delete push down, I left some comments there.
Btw, can the new `DeleteFromFilterOperation` extend the `SinkModifyOperation`? semantically, it is a dml operation for the sink table.

luoyuxia commented:
> can the new DeleteFromFilterOperation extend the SinkModifyOperation? 
I think we can make DeleteFromFilterOperation implement ModifyOperation since `SinkModifyOperation` contains much other things that won't be needed in `DeleteFromFilterOperation`.

luoyuxia commented:
@lincoln-lil Thanks for reviewing. I have addressed your comments about the first commits.
The test failure is about the error message donesn't match,
```
Expecting message to be:
 "Can't perform deletion for table default_catalog.default_database.t for the corresponding dynamic table sink haven't implement SupportsRowLevelDelete."
but was:
"Can't perform deletion for the table default_catalog.default_database.t for the corresponding dynamic table sink haven't implement org.apache.flink.table.connector.sink.abilities.SupportsRowLevelDelete."
```
Since it's introduced by the second commits,  I'll fix it while addressing comments about the second commits.

lincoln-lil reviewed (commented):
I've left some comments for the 2nd commit

lincoln-lil commented on flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/batch/sql/RowLevelDeleteTest.java:
```diff
@@ -0,0 +1,115 @@
+        util.verifyExplainInsert("DELETE FROM t where b = '123'");
+    }
+
+    private void createDeleteTable() {
```
nit: createTableForDelete

lincoln-lil reviewed (commented):
one more question for the 2nd row-level delete:
consider the delete query without a filter: "DELETE FROM t", does it work under 'REMAINING_ROWS' mode?
the generated plan contains an empty values source, so the sink will receive nothing till finish, should we add some notes for this to remind connector developers?

lincoln-lil commented:
> > can the new DeleteFromFilterOperation extend the SinkModifyOperation?
> 
> I think we can make DeleteFromFilterOperation implement ModifyOperation since `SinkModifyOperation` contains much other things that won't be needed in `DeleteFromFilterOperation`.

@luoyuxia In addition to the semantics that DeleteFromFilterOperation is also a SinkModifyOperation, another reason is that not all features supported by SinkModifyOperation are unused, such as dynamicOptions, WDYT?

luoyuxia commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/SinkModifyOperation.java:
```diff
@@ -43,6 +43,23 @@ public class SinkModifyOperation implements ModifyOperation {
     private final QueryOperation child;
     private final boolean overwrite;
     private final Map<String, String> dynamicOptions;
+    private final boolean isDelete;
```
Good suggestion.

luoyuxia commented:
> one more question for the 2nd row-level delete: consider the delete query without a filter: "DELETE FROM t", does it work under 'REMAINING_ROWS' mode? the generated plan contains an empty values source, so the sink will receive nothing till finish, should we add some notes for this to remind connector developers?

Yes, it works under 'REMAINING_ROWS' mode. That's true that the sink will recieive nothing. But I think we may not add a note for this case specially. IMO, if the mode is specified by 'REMAINING_ROWS' mode,  the devs should understand the behavior and implement it just like  `insert overwrite` which will delete the old data and write new data comming. 
It's nothing special when no data comming, it just is like it will skip write the data comming.

If needed, I think we can note it in the document.

luoyuxia commented:
@lincoln-lil Thanks for your reviewing. I have addressed your comments.

lincoln-lil commented on flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/batch/sql/RowLevelDeleteTest.java:
```diff
@@ -53,24 +57,37 @@ public RowLevelDeleteTest(SupportsRowLevelDelete.RowLevelDeleteMode deleteMode)
+        createTableForDelete();
+        util.verifyExplainInsert(
+                "DELETE FROM t",
+                JavaScalaConversionUtil.toScala(
```
nit: make it a member variable for reusing

lincoln-lil commented on flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/operations/DeletePushDownUtils.java:
```diff
@@ -0,0 +1,283 @@
+     * Get the resolve filter expressions from the {@code WHERE} clause in DELETE statement, return
+     * Optional.empty() if {@code WHERE} clause contains sub-query.
+     */
+    public static Optional<List<ResolvedExpression>> getResolveFilterExpressions(
```
nit: -> 'getResolvedFilterExpressions', and also update the comment

lincoln-lil replied:
nit: -> 'getResolvedFilterExpressions', and also update the comment

lincoln-lil commented on flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/batch/sql/RowLevelDeleteTest.java:
```diff
@@ -0,0 +1,138 @@
+        util.verifyExplainInsert(
+                "DELETE FROM t",
+                JavaScalaConversionUtil.toScala(
+                        Collections.singletonList(ExplainDetail.JSON_EXECUTION_PLAN)));
```
nit: make it a member variable for reusing

lincoln-lil reviewed (commented):
@luoyuxia the update looks good to me! Only two minor comments(one for each commit), and there's 1 failure case need to processed(update error msg)

luoyuxia commented:
@lincoln-lil Thanks for your reviewing. I have addressed your comments and test failure.
Sorry for the some wrong commit messages in the `3~5` commits.  They shouldn't be `[FLINK-30665]xxxx`., they should be `[FLINK-30662]xxxx`. Notice that just now.

lincoln-lil commented:
@luoyuxia never mind, we will finally squash these commits.

lincoln-lil reviewed (approved):
LGTM

lincoln-lil commented:
@luoyuxia the merged commits look like there are some changes and commits that are not aligned, so could you merge them again when you have time？
Chinese New Year is coming, wishing you a happy new year!

luoyuxia commented:
Thanks for the detail check. Sorry for missing some thing while merge these commits to two commit.  Now, I fix it, each commit should compile and pass test sucessfully.

Thank you for your blessing, Happy new year to you too.

klion26 commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/TableEnvironmentImpl.java:
```diff
@@ -847,6 +854,26 @@ public TableResultInternal executeInternal(List<ModifyOperation> operations) {
+                                        modifyType));
+                    }
+                    if (isStreamingMode) {
+                        throw new TableException(
```
Hi, @luoyuxia is there any document about this not supported statement(how to migrate them from previous versions)? and will we support this in the near future? thanks.

luoyuxia replied:
Thanks reach me out. What do you mean by saying `how to migrate them from previous versions`? Currently,it's by design that delete & update is not supported in stream mode  and the previous versioneven don't support delete & update statement. So, what is for migration?

klion26 replied:
@luoyuxia thanks for the reply, and sorry for the previous message. after another check, found that the problem is not caused here, will start discuss in the third project.
